src/analysis/lattice.h

```diff
@@ -19,32 +20,53 @@ constexpr bool has_getLeastUpperBound = std::
   is_invocable_r<void, decltype(&T::getLeastUpperBound), T, const T&>::value;
 template<typename T>
 constexpr bool has_isTop =
-  std::is_invocable_r<bool, decltype(T::isTop), const T&>::value;
+  std::is_invocable_r<bool, decltype(&T::isTop), T>::value;
 template<typename T>
 constexpr bool has_isBottom =
-  std::is_invocable_r<bool, decltype(T::isBottom), const T&>::value;
+  std::is_invocable_r<bool, decltype(&T::isBottom), T>::value;
 
 template<typename T>
 constexpr bool is_lattice =
   has_compare<T>&& has_getLeastUpperBound<T>&& has_isTop<T>&& has_isBottom<T>;
 
 // Represents a powerset lattice element (i.e. a set) as a bitvector. A true
 // means that an element is present in the set.
-template<size_t N> struct BitsetPowersetLattice {
-  std::bitset<N> value;
+class FinitePowersetLattice {
+  std::vector<bool> bitvector;
+
+  // Counts the number of elements in the set (i. e. number of trues in the
+  // bitvector).
+  size_t trues;
+
+  // Size of the top lattice element must be specified, as it determines the set
+  // of the bitvector.
+  FinitePowersetLattice(size_t size);
```

Comment:
I don't understand the comment. Is this function only used to create the top element, or something like that?

Reply:
It's a constructor. It actually creates the bottom element. But when I refer to the top element, I mean that the bitvector has to be the size of the top element (i.e. contain all possible set elements).

Reply:
Ah, ok, then I think the comment could maybe be reworded a little. It isn't the size of the top element that matters here. It is the size of the lattice (which happens to be the size of the top element).

That is, the comment could say, "when creating a lattice element we must pass in the size of the lattice".

Reply:
I would say it _is_ the size of the top element that matters here, if only because the top element happens to be the set that we are taking the powerset of. The "size of the lattice" is the number of elements in the powerset, which is not what we are referring to here.

Reply:
Ah, fair enough, the lattice is defined as the powerset, correct.